# Content and Task 4.7 - Deriving New Variables

Index:

##### 1.  Importing Data Frames and loading libraries 
##### 2.  If-Statements 
##### 2.1. If-Statements with User-Defined Functions
##### 2.2. If-Statements with the loc() Function
##### 2.3. If-Statements with For-Loops

##### 3.   Task 4.7
##### 3.1. Updating the lables
##### 3.2. Label the busiest periods of the day as "Most orders," "Average orders," and "Fewest orders
#####  3.3. Downloading data in to pickel


# 1.- Importing Data Frames and loading libraries

In [4]:
## Importing Data: 

import pandas as pd
import numpy as np
import os

path = r'C:\Users\cmfag\OneDrive\Desktop\Data Analyst Specialist\Achievement 4\4.3_orders_products\02 Data\Prepared Data'

In [6]:
#importing df_ords_prods_merge
# Verificar la data cargada!!

df_ords_prods_merge = pd.read_pickle(os.path.join(path, 'ords_prods_merge.pkl'))

df_ords_prods_merge.head()

NameError: name 'pd' is not defined

# 2.- If-Statements

###  2.1 If-Statements with User-Defined Functions

In [8]:
df = df_ords_prods_merge[:1000000]
#este subset incluye: everything from the beginning of the dataframe up to 1000000, es decir 1000000 rows.
df.shape

(1000000, 16)

In [9]:
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'


# esto quiere decir: 

#Rango de Bajo Costo (Low-range product):
   #Si el precio (prices) es menor o igual a 5, el producto se clasifica como "Low-range product".
   # Rango de Precio Medio (Mid-range product):
   #Si el precio está entre 5 y 15 (incluyendo 15), el producto se clasifica como "Mid-range product".
   # Rango de Alto Costo (High range):
   # Si el precio es mayor que 15, el producto se clasifica como "High range".
   # Datos Insuficientes (Not enough data):
   # Si el precio no cumple con ninguno de estos criterios, se etiqueta como "Not enough data" 
   #(esto es más un caso de cobertura para posibles datos faltantes o errores).

In [10]:
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\cmfag\AppData\Local\Temp\ipykernel_12380\2848743372.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [11]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    672525
Low-range product    314109
High range            12413
Not enough data         953
Name: count, dtype: int64

In [12]:
df['prices'].max()



99999.0

### !!!!!  Preguntar al tutor - en el content, me sale que no hay ningun producto en la categoria high-range, pero a mi me salen 12413 precios. 

## COMPARAR! Antes del If-Statements with the loc() Function sección 4.7

This function returns the maximum value within the “prices” column, which, as you’ll see here, is 14.8

## 2.2 If-Statements with the loc() Function

#### - esta opción te permite hacer las misma etiqueta/segmentación en toda la base de dato, sin necesidad de generar un subset para evitar
####   problemas de memoria.

In [15]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\cmfag\AppData\Local\Temp\ipykernel_12380\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [16]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [17]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [18]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    672525
Low-range product    314109
High range            12413
Not enough data         953
Name: count, dtype: int64

## 2.3 If-Statements with For-Loops



In [20]:
# Check column names in df_ords_prods_merge
print(df_ords_prods_merge.columns)


Index(['order_id', 'user_id', 'eval_set', 'order_number', 'order_dow',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'Unnamed: 0', 'product_name',
       'aisle_id', 'department_id', 'prices', 'merge_check'],
      dtype='object')


In [ ]:
# Check the result
df_ords_prods_merge[['order_hour_of_day', 'busiest_period_of_day']].head()

In [22]:
df_ords_prods_merge['busiest_day'] = result

In [23]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22437988
Busiest day        6209808
Least busy         3787263
Name: count, dtype: int64

In [24]:
df_ords_prods_merge.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

   Unnamed: 0                             product_name  aisle_id  \
0       195.0                                     Soda      77.0   
1     14084.0  Organic Unsweetened Vanilla Almond Milk      91.0   
2     12427.0                      Original Beef Jerky      23.0   
3     26089.0               Aged White Cheddar Popcorn      23.0   
4     26406.0         XL Pick-A-Size Paper Towel Rolls      54.0   

   department_id  prices merge_check       day_label     busiest_day  
0            7.0     9.0        both  Regularly busy  Regularly busy  
1           16.0    12.5        both  Regularly busy  Regularly busy  
2           19.0     4.4        both  Regularly busy  Regularly busy  
3           19.0     4.7        both  Regularly busy  Regularly busy  
4           17.0     1.0        both  Regularly busy  Regularly busy

In [25]:
df.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

   Unnamed: 0                             product_name  aisle_id  \
0       195.0                                     Soda      77.0   
1     14084.0  Organic Unsweetened Vanilla Almond Milk      91.0   
2     12427.0                      Original Beef Jerky      23.0   
3     26089.0               Aged White Cheddar Popcorn      23.0   
4     26406.0         XL Pick-A-Size Paper Towel Rolls      54.0   

   department_id  prices merge_check        price_range    price_range_loc  
0            7.0     9.0        both  Mid-range product  Mid-range product  
1           16.0    12.5        both  Mid-range product  Mid-range product  
2           19.0     4.4        both  Low-range product  Low-range product  
3           19.0     4.7        both  Low-range product  Low-range product  
4           17.0     1.0        both  Low-range product  Low-range product

### 3.- Task 4.7


### 3.1.- Updating the lables.

#### Creating a new column:
#### “Busiest day” to become “Busiest days” (plural) -  two busiest days of the week as opposed to the single busiest d
#### two slowest days
#### Warning: i will use df (1000 rows) since my laptop is quite slow and df_ords_prods_merge doesnt generates any reaction. 

In [51]:
# Step 1: Calculate frequency of orders by day

day_counts = df['order_dow'].value_counts().sort_values(ascending=False)

day_counts

order_dow
0    189677
1    169911
6    139623
5    133974
2    130568
3    118352
4    117895
Name: count, dtype: int64

In [53]:
# Identify the two busiest days (highest frequency)
busiest_days = day_counts.index[:2].tolist()

# Identify the two slowest days (lowest frequency)
slowest_days = day_counts.index[-2:].tolist()

In [55]:
busiest_days

[0, 1]

In [57]:
slowest_days

[3, 4]

In [60]:
# Step 2: Create a new column with the updated labels


def label_day(row):
    if row['order_dow'] in busiest_days:
        return "Busiest days"
    elif row['order_dow'] in slowest_days:
        return "Slowest days"
    else:
        return "Regularly busy"

In [62]:
df['day_activity_label'] = df.apply(label_day, axis=1)

df.head()

C:\Users\cmfag\AppData\Local\Temp\ipykernel_12380\4233308509.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['day_activity_label'] = df.apply(label_day, axis=1)


order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

   Unnamed: 0                             product_name  aisle_id  \
0       195.0                                     Soda      77.0   
1     14084.0  Organic Unsweetened Vanilla Almond Milk      91.0   
2     12427.0                      Original Beef Jerky      23.0   
3     26089.0               Aged White Cheddar Popcorn      23.0   
4     26406.0         XL Pick-A-Size Paper Towel Rolls      54.0   

   department_id  prices merge_check        price_range    price_range_loc  \
0            7.0     9.0        both  Mid-range product  Mid-range product   
1           16.0    12.5        both  Mid-range product  Mid-range product   
2           19.0     4.4        both  Low-range product  Low-range product   
3           19.0     4.7        both  Low-range product  Low-range product   
4           17.0     1.0        both  Low-range product  Low-range product   

  day_activity_label  
0     Regularly busy  
1     Regularly busy  
2     Regularly busy  
3     Regularly busy  
4     Regularly busy

In [64]:
# Calculate the counts for each label in 'day_activity_label'
label_counts = df['day_activity_label'].value_counts()

# Display the counts
print(label_counts)

day_activity_label
Regularly busy    404165
Busiest days      359588
Slowest days      236247
Name: count, dtype: int64


#### Comments: everything looks well, any specific comment, the counts is right and the overview out of the head is to the range. 

## 3.2.- Label the busiest periods of the day as "Most orders," "Average orders," and "Fewest orders

In [84]:
# Step 1: Calculate order frequency by hour

hour_counts = df['order_hour_of_day'].value_counts().sort_index()
hour_counts

order_hour_of_day
0      5968
1      2938
2      1867
3      1482
4      1667
5      2684
6      9534
7     27795
8     53841
9     77748
10    86511
11    84881
12    80836
13    81562
14    81395
15    81136
16    76747
17    65223
18    51360
19    39414
20    29656
21    24552
22    19242
23    11961
Name: count, dtype: int64

In [86]:
# Step 2: Determine thresholds for "Most orders," "Average orders," and "Fewest orders"

# Using quantiles for categorization

# top 1/3 for "Most orders"

most_threshold = hour_counts.quantile(0.66)  

# bottom 1/3 for "Fewest orders"
fewest_threshold = hour_counts.quantile(0.33) 



In [88]:
most_threshold

67297.32000000002

In [90]:
fewest_threshold

16256.790000000005

In [102]:
# Step 3: Use .loc to label the 'busiest_period_of_day' based on thresholds
df['busiest_period_of_day'] = 'Average orders'  # Default value

df.loc[df['order_hour_of_day'] <= fewest_threshold, 'busiest_period_of_day'] = 'Fewest orders'
df.loc[df['order_hour_of_day'] > most_threshold, 'busiest_period_of_day'] = 'Most orders'

C:\Users\cmfag\AppData\Local\Temp\ipykernel_12380\51233542.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['busiest_period_of_day'] = 'Average orders'  # Default value


In [116]:
# Check the result
df[['order_hour_of_day', 'busiest_period_of_day']].head()

order_hour_of_day busiest_period_of_day
0                  8         Fewest orders
1                  8         Fewest orders
2                  8         Fewest orders
3                  8         Fewest orders
4                  8         Fewest orders

In [112]:
label_counts_houw = df['busiest_period_of_day'].value_counts()

# Display the counts
print(label_counts)


day_activity_label
Regularly busy    404165
Busiest days      359588
Slowest days      236247
Name: count, dtype: int64


#### Comment: count is right. 

## 3.3 Export the dataframe

In [183]:
df_ords_prods_merge.to_pickle(os.path.join(path, 'df_ords_prods_merge.pkl'))